### DISCUSSION TITLE
Reflection/metadata of cue fields for generating terraform variables.tf file.

### DISCUSSION THREAD

**Discussion by pj** [cue-lang/cue#1865] [Q&A]
Hi there, I might have missed this, but is there some kind of reflection or metadata system for cue? i.e.:

```cue
a: {
  b: string | *"Hello, World!"
  c: int | *42
}
```

Could be transformed into an AST representation like the following, without needing to write go code?

```cue
a: {
 type: "struct"
 fields: {
    b: {type: "string", default: "Hello, World!"}
    c: {type: "int", default: 42}
  }
}
```

Context is that I would like to generate a terraform variables schema files that look like:

```terraform
variable "a" {
   type = string
   description = "A useful variable"
  default = "Hello, world!"
}

variable "b" {
   type = number
  description = "Life, universe etc"
  default = 42
}

```

**Comment by pj** [comment]
I worked around this by writing some go code to search for definitions with a specific attribute:
```cue
#GlobalTerraformVariables: {
	global_project_id:          #Main.global.project.id @generate_schema() 
	global_project_name:        #Main.global.project.name
	artifact_repository_region: #Main.global.artifact_repository_region
} @generate_terraform_variables(filename="../global/variables.tf") 
```

If I had a bit more time and better examples it would be good to expand it and make a tool out of it, so that it's easy to generate metadata in cue format, to be consumed in other files.

I put the code in a gist: https://gist.github.com/pj/687745e85f4d5ed104b292cebd1cd426

**Comment by verdverm** [comment]
https://github.com/cue-lang/cue/discussions/1889 is related and I wrote many thoughts there that I would have written here. I am maintaining my thoughts about CUE -> type in languages here: https://docs.hofstadter.io/code-generation/type-and-structs/

To add for the context here...

- attributes are a bit finicky, and I'm moving away from them for type / field metadata. I primarily use them for mapping to command line args / flags for the tool I work on (https://github.com/hofstadter-io/hof)
- I now prefer DSL / real CUE metadata for these. You get the CUE evaluators correctness applying to them
- Attributes are intentionally not part of the evaluation process. So you lose many of the benefits of writing in CUE, and you end up having to write the validation in Go, which as you can imagine is a pain when you have CUE's power so close at hand.

The general strategy I am using is to turn the CUE into concrete data and feed this into text/template like you have. We built a purposeful tool for this process which makes it easy to create the DSL (CUE schemas) and templates without modifying the tool. It runs much faster, has module support for composition, and much much more. I've been looking at how to turn the incomplete / abstract CUE into the DSL expected by a generator (template). My latest experiment is here:

- https://github.com/hofstadter-io/hofmod-types/blob/main/generators/base/gen.cue#L14 (applied on line 30 in the same file)
- https://github.com/hofstadter-io/hofmod-types/blob/main/generators/go/gen.cue#L12 (injected here)
- https://github.com/hofstadter-io/hof/blob/_dev/schema/dm/enrichers/go.cue

The main gap right now is that determining the specific type from incomplete CUE has too many edge cases to create a solid UX/DX right now. This is something I am hopeful will improve with time.

**Comment by pj** [comment]
Thanks for the response!

I guess there are good reasons, but it does seem a bit unfortunate that attributes aren't real cue. I do like the syntax of them, oh well. 

In terms of types, a lot of other languages are pretty dumb when it comes to their "type" system or schema (like terraform variables files). I understand there are more complicated scenarios where you might need something more powerful, but there are a lot of cases where something basic would work and not require a whole other DSL to do things. 

If you have to build a language within a language and use templates to generate strings just to convert a cue "string" field to a terraform variables "string" field, I struggle a bit too see what I'm gaining over the alternatives?

Thanks again.

**Reply by verdverm** [reply]
I would say that any way about it, you are transforming from one codec to another, essentially writing a transpiler, even if it is a function. You have to write code to do this at some point. It would put a lot of burden on the CUE team to have to maintain every target language someone might want to turn CUE into. So there are a couple of options (1) write a Go package and tool (2) use text/template. Both look like string interpolation to me.

As for what you are gaining over alternatives, the logic and correctness that CUE offers generally would be (is) my response.

I've been just writing TF as CUE directly, exporting it to `resource.tf.json` files and running TF against them directly. So I don't have a need for transpile to TF files. If you are only targeting TF, I would highly recommend this path.

**Reply by pj** [reply]
Writing everything in cue makes it harder to copy and paste from stackoverflow, how will I survive? 😂. 

More seriously, do you know of any examples of tools that are external to the cue repository? It's a little bit unclear how you would contribute interface code and similar sorts of things to the cue eco system.

**Reply by verdverm** [reply]
I am the creator `hof` (https://docs.hofstadter.io | https://github.com/hofstadter-io/hof) which aims to enable an ecosystem around the concepts I have been describing. The input schemas (DSL) and templates (generator modules) are all external to the tool.

Is that the type of thing you are talking about or more of a tool with specific codecs in mind?

**Comment by verdverm** [comment]
Regarding a reflection feature, I am highly interested in this as well. It would be great to make it easier to turn "abstract" CUE into more structured DSL. We should definitely raise a feature request around this, I'm sure the CUE team would be interested to support something like this in a way that makes sense, taking consideration of how CUE works. (in particular the underlying theory / spec)

**Reply by pj** [reply]
I was wondering about a tool that dumps out metadata as a cue file, so it can be consumed by subsequent steps. That way you don't have to modify the core language.

**Reply by verdverm** [reply]
I have some helper funcs here you can use or copy:

- https://github.com/hofstadter-io/hof/blob/_dev/lib/cuetils/attrs.go
- https://github.com/hofstadter-io/hof/blob/_dev/lib/gen/runtime.go#L201

You'll of course have to parse & process the part between the parens as needed. Note that if the same attribute appears on a single field more than once, they are not combined because it is too application specific for CUE to decide how that should be done.

**Comment by eadlam** [comment]
In a limited sense, cue almost supports this. Anything that's a definition can be exported to OpenAPI, which is effectively an AST of the types  and values. OpenAPI is json, and since cue is a superset of json, OpenAPI is cue. So a cue spec can generate a quasi AST of itself in cue.
